### PR TITLE
Purge the root vendor directory

### PR DESCRIPTION
**Pull request by timflannagan**
Update the repository layout and avoid commiting the root vendor directory into source. Update the root .gitignore and ignore the vendor directory. For any CI/CD (or utility-esq Makefile targets) avoid referencing the vendor equivalent through `go run ...` and instead use the go-get-tool from the operator-sdk/kubebuilder world) and pull down that dependency at runtime.

### CODE REVIEW

**Comment by timflannagan** [comment]
I realize we need to also bump this repository to Go 1.17.

File: tools.go
Comment (joelanford) [inline tools.go:7]: Are we able to use go-get-tool for go-bindata as well?

File: Makefile
Comment (joelanford) [inline Makefile:102]: Just as a heads up.. this is fine for now, but it's going to break when we bump to Go 1.18. See https://github.com/operator-framework/operator-sdk/issues/5481

Comment (timflannagan) [reply Makefile:102]: That's a good call out - I'm personally not super worried if we're committed to tracking this here as an issue (or ideally in jira ticket) as it looks like a fairly straight forward change, and removing the vendor directory is a step in the right directory (both in the context of upstream, and especially downstream) imo.

**Comment by timflannagan** [comment]
I'd like to resolve that go-apidiff before merging this to avoid putting this repository in a state where we break that action, and potentially evaluate purging the root tools.go file.

/hold

**Comment by timflannagan** [comment]
Ah, it looks like we don't have the workflow_dispatch event trigger enabled for this repository. Opened a quick PR (https://github.com/operator-framework/api/pull/214) for enabling that for the two workflow definitions.

File: tools.go
Comment (timflannagan) [reply tools.go:7]: Surprisingly, this was a bit more complicated that I had anticipated due to how bloated the `manifests` target is right now, and exporting the PATH variable to point to this repositories' root `./bin` directory before running `go generate ...` wasn't working, and that command was always searching for packages in the vendor directory, despite no GOFLAGS being set in my local environment.

It's possible I'm missing something obvious, but given the o-f/api recent changes that contain massive changelogs due to dependency bumps, maybe we can tackle this in a follow-up, and/or generalize the value we see in the tools.go file going forward throughout the OLM-controller repositories?

**Comment by fgiloux** [comment]
I have seen how annoying the vendor directory is during PR. That said there are good reasons for keeping a copy of the dependencies. A possible mitigation could be to have the vendor directory stored in a separate repository when a new release is cut. What do you think?

**Comment by timflannagan** [comment]
I think the value of the vendor directory has somewhat deteriorated as modules become more stable, and my preference would be to remove the vendor directory from upstream repositories, while maintaining the vendor directory in any downstream environments. That should serve as a nice balance where we can rely on dependencies being available for downstream builds, but help improve the upstream contributor/reviewer experience. Thoughts?

**Comment by fgiloux** [comment]
> to remove the vendor directory from upstream repositories, while maintaining the vendor directory in any downstream 

Yes that sounds good to me. My main concern was indeed when there is a need to patch a downstream release and the dependency version is nowhere to be found.

**Review by dinhxuanvu — APPROVED** [review]
/approve

File: Makefile
Comment (timflannagan) [reply Makefile:102]: I'll create an ticket in this repository once this PR merges to track this.

**Pull request merged**
